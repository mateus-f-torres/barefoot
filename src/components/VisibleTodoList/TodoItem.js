//@flow
import * as React from 'react';
import styled from 'styled-components';
import type {Todo} from 'Types/props.js';
import Button from 'Components/common/Button';
import trash from 'Assets/images/trash.png';

const MAIN_TEXT= '#1b1b1e';

const ListItem = styled.li`
  display: flex;
  background-color: darkgrey;
  color: ${MAIN_TEXT};
  justify-content: space-between;
  align-items: center;
  height: 45px;
  line-height: 45px;
  &:nth-child(2n) { 
    background-color: white;
    color: #666;
  }
`;

const ListItemText = styled.p`
  display: inline-block;
  width: 100%;
  padding-left: 10px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  text-decoration: ${(props) => props.done ? 'line-through' : 'none'};
  &:hover {
    cursor: pointer;
  }
`;

const ListItemIcon = styled.img`
  width: 30px;
  heigth: 30px;
  position: relative;
  bottom: 5px;
`;

type Props = Todo & {
  toggleTodo: (SyntheticEvent<HTMLParagraphElement>) => void,
  deleteTodo: (SyntheticEvent<HTMLButtonElement>) => void,
};

const TodoItem = (props: Props) => (
  <ListItem id={props.id}>
    <ListItemText
      done={props.completed ? true : false}
      data-test="todo-text"
      onClick={props.toggleTodo}>
      {props.text}
    </ListItemText>
    <Button
      data-test="todo-delete"
      onClick={props.deleteTodo}>
      <ListItemIcon src={trash}/>
    </Button>
  </ListItem>
);

export default TodoItem;
