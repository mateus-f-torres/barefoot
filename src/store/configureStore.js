//@flow
import {createStore} from 'redux';
import rootReducer from '../reducers/index.js';
import type {Store} from '../types/store.js';

export default function configureStore(): Store {
  return createStore(rootReducer);
}
