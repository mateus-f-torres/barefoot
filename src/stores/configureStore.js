//@flow
import {createStore} from 'redux';
import rootReducer from './index.js';
import type {Store} from 'Types/store.js';

export default function configureStore(): Store {
  return createStore(rootReducer);
}
