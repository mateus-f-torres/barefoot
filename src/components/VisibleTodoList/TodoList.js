//@flow
import * as React from 'react';
import styled from 'styled-components';
import type {Todo} from 'Types/props.js';
import TodoItem from './TodoItem.js';
import Button from 'Components/common/Button';

const MAIN_TEXT= '#1b1b1e';
const HEADER_BACK= '#1b3039';

const Form = styled.form`
  display: flex;
  flex-direction: column;
`;

const Input = styled.input`
  font-size: 18px;
  color: ${MAIN_TEXT};
  background-color: #f7f7f7;
  width: 100%;
  padding: 13px 13px 13px 20px;
  box-sizing: border-box;
  border: 3px solid rgba(0,0,0,0);
  &:focus {
    border: 3px solid ${HEADER_BACK};
    outline: none;
  }
`;

const List = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
`;

type Props = {
  todos: Array<Todo>,
  addTodo: (string) => void,
  toggleTodo: (number) => number,
  removeTodo: (number) => number,
};

class TodoList extends React.Component<Props> {
  handleSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();
    e.stopPropagation();
    if (e.currentTarget) {
      // $FlowFixMe, 'maybe-there' flow error
      let text = e.currentTarget.childNodes[0].value;
      if (!text.trim()) return;
      this.props.addTodo(text);
      // $FlowFixMe, 'maybe-there' flow error
      e.currentTarget.childNodes[0].value = '';
    }
  }

  toggleTodo = (e: SyntheticEvent<HTMLParagraphElement>) => {
    e.preventDefault();
    e.stopPropagation();
    // $FlowFixMe, 'maybe-there' flow error
    let id = e.currentTarget.parentNode.id;
    id = Number.parseInt(id);
    this.props.toggleTodo(id);
  }

  deleteTodo = (e: SyntheticEvent<HTMLButtonElement>) => {
    e.preventDefault();
    e.stopPropagation();
    // $FlowFixMe, 'maybe-there' flow error
    let id = e.currentTarget.parentNode.id;
    id = Number.parseInt(id);
    this.props.removeTodo(id);
  }

  render() {
    return (
      <Form
        data-test="todo-submit"
        onSubmit={this.handleSubmit}>
        <Input
          data-test="todo-input"
          type="text"
          placeholder="Add todo" />
        <Button type="submit" />
        <List data-test="todo-list">
          {
            this.props.todos.map((todo) => (
              <TodoItem
                key={todo.id}
                toggleTodo={this.toggleTodo}
                deleteTodo={this.deleteTodo}
                {...todo}
              />
            ))
          }
        </List>
      </Form>
    );
  }
}

export default TodoList;

