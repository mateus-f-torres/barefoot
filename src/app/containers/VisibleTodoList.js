//@flow
import {connect} from 'react-redux';
import {addTodo, toggleTodo, removeTodo} from '../actions';
import TodoList from '../components/TodoList.jsx';

const mapStateToProps = (state) => {
  return {
    todos: state.todos,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    addTodo: (text) => {
      dispatch(addTodo(text));
    },
    toggleTodo: (id) => {
      dispatch(toggleTodo(id));
    },
    removeTodo: (id) => {
      dispatch(removeTodo(id));
    },
  };
};

const VisibleTodoList = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default VisibleTodoList;
