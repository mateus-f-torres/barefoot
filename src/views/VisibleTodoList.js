//@flow
import {connect} from 'react-redux';
import {addTodo, toggleTodo, removeTodo} from 'Store/todos/actions.js';
import TodoList from './VisibleTodoList/TodoList.js';
import type {State} from 'Types/state.js';

const mapStateToProps = (state: State) => {
  return {
    todos: state.todos,
  };
};

const mapDispatchToProps = (dispatch: *) => {
  return {
    addTodo: (text) => {
      dispatch(addTodo(text));
    },
    toggleTodo: (id) => {
      dispatch(toggleTodo(id));
    },
    removeTodo: (id) => {
      dispatch(removeTodo(id));
    },
  };
};

const VisibleTodoList = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default VisibleTodoList;
