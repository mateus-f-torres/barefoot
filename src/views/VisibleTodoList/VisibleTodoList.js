//@flow
import {connect} from 'react-redux';
import {addTodo, toggleTodo, removeTodo} from 'Stores/todos/actions';
import type {State} from 'Types/state';

import TodoList from './TodoList';

function mapStateToProps(state: State) {
  return {
    todos: state.todos,
  };
}

function mapDispatchToProps(dispatch: *) {
  return {
    addTodo(text) {
      dispatch(addTodo(text));
    },
    toggleTodo(id) {
      dispatch(toggleTodo(id));
    },
    removeTodo(id) {
      dispatch(removeTodo(id));
    },
  };
}

const VisibleTodoList = connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);

export default VisibleTodoList;
