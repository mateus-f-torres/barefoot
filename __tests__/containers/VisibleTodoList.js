import React from 'react';
import {shallow} from 'enzyme';
import configureStore from 'redux-mock-store';
import VisibleTodoList from '../../src/containers/VisibleTodoList.js';

const mockState = {
  todos: [
    {
      id: 0,
      text: 'mock1',
      completed: false,
    },
    {
      id: 1,
      text: 'mock2',
      completed: true,
    },
    {
      id: 2,
      text: 'mock3',
      completed: false,
    },
  ],
};

const mockStore = configureStore();
const store = mockStore(mockState);

describe('VisibleTodoList', () => {
  beforeEach(() => {
    store.clearActions();
  });

  const container = shallow(<VisibleTodoList />, {context: {store}});
  // console.log(container.debug()); for more info

  describe('Unit Test', () => {
    test('render TodoList component', () => {
      expect(typeof container.find('TodoList')).toBe('object');
    });

    test('receive store.todos', () => {
      expect(container.find('TodoList')
        .prop('todos')).toEqual(mockState.todos);
    });

    test('render all todos', () => {
      expect(container.dive().find('[data-test="todo-list"]')
        .children().length).toEqual(mockState.todos.length);
    });

    test('dispatch correct addTodo', () => {
      const mockEvent = {
        preventDefault() {},
        stopPropagation() {},
        currentTarget: {childNodes: [{value: 'Hello World'}]},
      };

      const expectedActions = [
        {
          'type': 'ADD_TODO',
          'payload': 'Hello World',
        },
      ];

      container.dive().find('[data-test="todo-submit"]')
        .simulate('submit', mockEvent);

      expect(store.getActions()).toEqual(expectedActions);
    });

    test('dont dispatch addTodo without value', () => {
      const mockEvent = {
        preventDefault() {},
        stopPropagation() {},
        currentTarget: {childNodes: [{value: '    '}]},
      };

      container.dive().find('[data-test="todo-submit"]')
        .simulate('submit', mockEvent);

      expect(store.getActions()).toEqual([]);
    });

    test('dont dispatch addTodo without currentTarget', () => {
      const mockEvent = {
        preventDefault() {},
        stopPropagation() {},
      };

      container.dive().find('[data-test="todo-submit"]')
        .simulate('submit', mockEvent);

      expect(store.getActions()).toEqual([]);
    });

    test('reset input value after addTodo', () => {
      const mockEvent = {
        preventDefault() {},
        stopPropagation() {},
        currentTarget: {childNodes: [{value: 'Hello World'}]},
      };

      container.dive().find('[data-test="todo-submit"]')
        .simulate('submit', mockEvent);

      expect(mockEvent.currentTarget.childNodes[0].value).toBe('');
    });

    test('dispatch correct toggleTodo', () => {
      const mockEvent = {
        preventDefault() {},
        stopPropagation() {},
        currentTarget: {parentNode: {id: '1'}},
      };

      const expectedActions = [
        {
          'type': 'TOGGLE_TODO',
          'payload': 1,
        },
      ];

      container.dive().find('[data-test="todo-list"]')
        .childAt(1).dive().find('p').simulate('click', mockEvent);

      expect(store.getActions()).toEqual(expectedActions);
    });

    test('dispatch correct removeTodo', () => {
      const mockEvent = {
        preventDefault() {},
        stopPropagation() {},
        currentTarget: {parentNode: {id: '2'}},
      };

      const expectedActions = [
        {
          'type': 'REMOVE_TODO',
          'payload': 2,
        },
      ];

      container.dive().find('[data-test="todo-list"]')
        .childAt(1).dive().find('button').simulate('click', mockEvent);

      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  describe('Snapshot Test', () => {
    test('matches current production snapshot', () => {
      expect(container).toMatchSnapshot();
    });
  });
});
