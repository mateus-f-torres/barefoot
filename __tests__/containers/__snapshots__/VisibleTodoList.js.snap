// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisibleTodoList Snapshot Test matches current production snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(TodoList)>
    <TodoList
      addTodo={[Function]}
      removeTodo={[Function]}
      todos={
        Array [
          Object {
            "completed": false,
            "id": 0,
            "text": "mock1",
          },
          Object {
            "completed": true,
            "id": 1,
            "text": "mock2",
          },
          Object {
            "completed": false,
            "id": 2,
            "text": "mock3",
          },
        ]
      }
      toggleTodo={[Function]}
    >
      <form
        data-test="todo-submit"
        onSubmit={[Function]}
      >
        <input
          data-test="todo-input"
          placeholder="Add todo"
          type="text"
        />
        <button
          type="submit"
        />
        <ul
          data-test="todo-list"
        >
          <TodoItem
            completed={false}
            deleteTodo={[Function]}
            id={0}
            key="0"
            text="mock1"
            toggleTodo={[Function]}
          >
            <li
              id={0}
            >
              <p
                className={null}
                data-test="todo-text"
                onClick={[Function]}
              >
                mock1
              </p>
              <button
                data-test="todo-delete"
                onClick={[Function]}
              >
                delete
              </button>
            </li>
          </TodoItem>
          <TodoItem
            completed={true}
            deleteTodo={[Function]}
            id={1}
            key="1"
            text="mock2"
            toggleTodo={[Function]}
          >
            <li
              id={1}
            >
              <p
                className="done"
                data-test="todo-text"
                onClick={[Function]}
              >
                mock2
              </p>
              <button
                data-test="todo-delete"
                onClick={[Function]}
              >
                delete
              </button>
            </li>
          </TodoItem>
          <TodoItem
            completed={false}
            deleteTodo={[Function]}
            id={2}
            key="2"
            text="mock3"
            toggleTodo={[Function]}
          >
            <li
              id={2}
            >
              <p
                className={null}
                data-test="todo-text"
                onClick={[Function]}
              >
                mock3
              </p>
              <button
                data-test="todo-delete"
                onClick={[Function]}
              >
                delete
              </button>
            </li>
          </TodoItem>
        </ul>
      </form>
    </TodoList>
  </Connect(TodoList)>
</Provider>
`;
